### DISCUSSION TITLE
LOL how in the world js nanoexpress have better performance than Go or Go Fiber

### DISCUSSION THREAD

**Discussion by kamalshkeir** [the-benchmarker/web-frameworks#4520] [Q&A]
For helloworld project, i did the test on my machine and everytime i get the same result, better performance for GoFiber.
GoFiber handle 77k req/s 
nanoexpress handle 50k req/s

**Comment by waghanza** [comment]
Hello @kamalshkeir,

Thanks for concern.

First of all, I will say that this project is still in **beta** stage.

I understand that you are concern about `fiber` vs `nanoexpress`. I'll try to explain the differences in 2 points :
- as you noticed https://web-frameworks-benchmark.netlify.app/result?f=nanoexpress,fiber are not corresponding to any above figures -> the idea is to have a comparison in an architecture corresponding to what **IS** server (e.g : **production** server / vms / pods)
- we are using `docker` -> I'm not sure we can correlate any real-word figure with **local** `docker`
- are we sure we use the same version / go compilation options / os ?

Regards,

**Reply by kamalshkeir** [reply]
I was shoked to see a dynamic language like Js beat a compiled language like Go, i have test my Go server against .Net , Django, Ruby..., never seen any server that can handle 70k req/s like Go do.
Anw i understand how hard to simulate a real world benchmark that work the same on every machine.
btw i was using go 1.16.6 on windows machine localy vs nanoexpress on the same machine locally without docker for my benchmarks.
Thank u for ur answer

**Reply by waghanza** [reply]
Very hard explain those results differences, but what I can say is :
+ networks are **NOT** handled the same way on `unix` and `windows` (I'm not sure their is any value to explain those gaps)
+ a interpreted language could be faster for a particular benchmark -> for example, **most** of stack use advanced techniques (`jit`, `low lever wrappers` ....) that infirm this assertion

**Comment by re-thc** [comment]
The test isn't really testing `javascript` vs `go`. It only tests the web server component.

It makes sense that `nanoexpress` has comparable performance to `fiber`. Besides a very minor part of the code, `nanoexpress` uses heavily optimized c/c++ in the form of `uWebSockets`, which would be faster than Go.

If you're interested in making a point you can try Deno (based also on javascript / v8) with its new Rust based Http server and that might be even faster :).

**Reply by waghanza** [reply]
> The test isn't really testing javascript vs go. It only tests the web server component.

indeed, comparing languages could only be made using the same kind of component, `here` (server-side) frameworks.

> It makes sense that nanoexpress has comparable performance to fiber. Besides a very minor part of the code, nanoexpress uses heavily optimized c/c++ in the form of uWebSockets, which would be faster than Go.

Right, some frameworks uses some low level components (mostly to deal with network) such as  https://www.swoole.co.uk in `php`, https://github.com/jonashaag/bjoern in `python`, https://deno.land/ in `javascript` ...

The point is that any language **COULD** (in theory) be faster than expected. It depends mainly on code design (async/sync, complexity, abstraction ...)

**Reply by waghanza** [reply]
Please not however, that the actual results are still in beta. I mean no real network is used (only loopback interface) and then results probably would vary a lot after running on a real-life network

**Comment by bangbaew** [comment]
![image](https://user-images.githubusercontent.com/29559559/154520187-ecc30c8e-57b6-4836-87e3-4a65b9818dd3.png)

I've tried load testing using K6 (64 VUs, 10s), from left to right: Nest.js Fastify, Gofiber, Nanoexpress
and Nanoexpress gave me impressive results. 
The syntax is almost identical to Express, just change from `app.delete` method to `app.del`,
I've simply changed only `require('express')` to `require('nanoexpress')` and it runs normally.

<img width="399" alt="image" src="https://user-images.githubusercontent.com/29559559/154524574-ccc47bc1-8ac2-4f1d-9dad-810b288adc79.png">
<img width="396" alt="image" src="https://user-images.githubusercontent.com/29559559/154526022-b735e7d5-92d7-44d8-853c-97d0157f513f.png">


For the second, third and fourth try, Nanoexpress had a little drop in performance but still beat Gofiber

**Comment by sharpninja** [comment]
Are the official test result run in GitHub Actions?

**Reply by waghanza** [reply]
nop. their are running on a private workstation. the choice was made since this workstation has closer specs facing some cloud servers (the idea behind that is to run a a real cloud when this project will be mature and in production)

**Comment by waghanza** [comment]
I'll close this discussion as we can not take any conclusion in this results as for now.

I've opened a discussion https://github.com/the-benchmarker/web-frameworks/discussions/8093 to enhance our test suite
